<h1><center>Lab Assignment 6</center></h1>

<center> DUE: <font color = 'blue'> Tuesday, November 14th by 3:00 PM (EST)</font> via Canvas </center>

Please submit:

1) a PDF with all code in-line, generated from Jupytor Notebook
  - If there is an error when you generate pdf, please refer this.
  [4 ways to export Jupyter Notebook as PDF](https://mljar.com/blog/jupyter-notebook-pdf/#:~:text=The%20Jupyter%20Notebook%20has%20an,not%20visible%20in%20the%20screenshot))
  - <font color = red>If you submit your result as ipynb extension, you will deduct points.</font>

<h3><center>Asad Hashmi</center></h3>

**Round your answers to two decimal places, and include all code and figures.**

### Problem Description
A study was performed to investigate new automobile purchases. A sample of 20 families was selected. Each family was surveyed to determine their total family income ($x_1$) and the age of their oldest vehicle ($x_2$). A follow-up survey was conducted six months later to determine if they had purchased a new vehicle during that time period ($y = 1$ indicates YES and $y = 0$ indicates NO.) Use the dataset `lab6_data.csv`. 

In [2]:
import numpy as np                    #to use matrix operations
import pandas as pd                   #to handle data
import matplotlib.pyplot as plt       #to draw plot
import statsmodels.api as sm          #to use statistical model
import statsmodels.formula.api as smf #to use statistical model

### Data loading

In [3]:
# read in file
df = pd.read_csv("lab6_data.csv")
df.head()

x1  x2  y
0  45000   2  0
1  40000   4  0
2  60000   3  1
3  50000   2  1
4  55000   2  0

### (a) (3 points)

**Fit a *multiple linear regression* with both covariates income ($x_1$) and age ($x_2$). Report your regression equation, the R-squared value, and the adjusted R-squared value. (For the regression equation, please round coefficient estimates to *five* decimal places.)**

In [4]:
lm_model = smf.ols(
        formula = 'y ~ x1 + x2',data=df).fit()
lm_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.263
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     3.031
Date:                Sun, 12 Nov 2023   Prob (F-statistic):             0.0748
Time:                        17:54:44   Log-Likelihood:                -11.466
No. Observations:                  20   AIC:                             28.93
Df Residuals:                      17   BIC:                             31.92
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.4966      0.706     -0.703      0.491      -1.987       0.993
x1          9.485e-06   1.09e-05      0.867      0.398   -1.36e-05    3.26e-05
x2             0.1462      0.064      2.300      0.034       0.012       0.280
==============================================================================
Omnibus:                        5.892   Durbin-Watson:                   1.800
Prob(Omnibus):                  0.053   Jarque-Bera (JB):                1.769
Skew:                           0.213   Prob(JB):                        0.413
Kurtosis:                       1.607   Cond. No.                     3.33e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.33e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The regression equation is:

y = -0.49660 + 0.000009485x1 + 0.14620x2, The R-squared value is 0.263, The adjusted R-squared value is 0.176


### (b) (1 points)

**Use your output from part (a) to conduct a model utility test using a significance level of $\alpha=0.05$ to decide if the model is significant. State your conclusion and how you obtained it.**

**Answer**

The Prob (F-statistic) = 0.0748 and our chosen significance level (𝛼) is 0.05. Therefore, we fail to reject the null hypothesis and we cannot conclude that the model is significant at the 0.05 level.

### (c) (5 points)

**Fit a multiple logistic regression model to the data.**

1. **Report the regression equation. (Round coefficient estimates to *five* decimal places).**

In [6]:
multiple_logit_model = smf.logit(formula = 'y ~ x1 + x2', data = df).fit()
multiple_logit_model.summary()

Optimization terminated successfully.
         Current function value: 0.526785
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                   20
Model:                          Logit   Df Residuals:                       17
Method:                           MLE   Df Model:                            2
Date:                Sun, 12 Nov 2023   Pseudo R-squ.:                  0.2400
Time:                        18:11:18   Log-Likelihood:                -10.536
converged:                       True   LL-Null:                       -13.863
Covariance Type:            nonrobust   LLR p-value:                   0.03589
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -7.0138      4.633     -1.514      0.130     -16.094       2.066
x1          7.354e-05   6.33e-05      1.162      0.245   -5.05e-05       0.000
x2             0.9831      0.523      1.879      0.060      -0.042       2.009
==============================================================================
"""

**Answer**
The logistic regression equation based on the output provided would be:
y = -7.01380 + 0.00007354x1 + 0.98310x2

2. **State the covariates that are significant at a significance level of $\alpha=0.1$.**

**Answer**

The p-value for x2 is 0.060, which is less than 𝛼=0.1, so x2 is statistically significant at a significance level of  𝛼=0.1.

### (d) (2 points)

1. **Obtain the odds ratio for each of the covariates.**

In [8]:
np.exp(multiple_logit_model.params)

Intercept    0.000899
x1           1.000074
x2           2.672802
dtype: float64

**Answer**
x1 = 1.000074
x2 = 2.672802


2. **Interpret the meaning of the odds ratios in words.**

**Answer**

The odds ratio for a variable represents the multiplicative change in the odds for a 1-unit increase in the variable, holding all other variables constant.

x1: This corresponds to an increase of odds by a factor of 1.00007354 for each one-unit increase in x1, holding x2 constant. However, because the odds ratio is so close to 1, the effect on the odds is very small and may not be considered meaningful.

x2: This corresponds to an increase of odds by a factor of approximately 2.67265 for each one-unit increase in x2, holding x1 constant. This means, all else being equal, for each one-unit increase in X2, the odds that 'y' occurs (i.e., y=1) are roughly 2.67 times higher.

### (e) (1 points) 

**What is the estimated probability that a family with an income of $45,000 and a car that is 5 years old will purchase a new vehicle in the next six months?**

In [9]:
# Define coefficients
coef_intercept = -7.01380
coef_x1 = 0.00007354
coef_x2 = 0.98310

# Define variables
x1 = 45000  # Income
x2 = 5  # Car age

# Compute log-odds
log_odds = coef_intercept + coef_x1 * x1 + coef_x2 * x2

# Compute probability with logistic function
prob = 1 / (1 + np.exp(-log_odds))

print("The estimated probability that a family with an income of $45,000 and a car that is 5 years old will purchase a new vehicle in the next six months is {:.2f}".format(prob))


The estimated probability that a family with an income of $45,000 and a car that is 5 years old will purchase a new vehicle in the next six months is 0.77


**Answer**
0.77
